openapi3/schema_merge_test.go

```diff
@@ -0,0 +1,1273 @@
+package openapi3
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// identical numeric types are merged successfully
+func TestMerge_TypeNumeric(t *testing.T) {
+	merged, err := Merge(Schema{
+		AllOf: SchemaRefs{
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "number",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "number",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, "number", merged.Properties["prop1"].Value.Type)
+
+	merged, err = Merge(Schema{
+		AllOf: SchemaRefs{
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "integer",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "integer",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, "integer", merged.Properties["prop1"].Value.Type)
+}
+
+// Conflicting numeric types are merged successfully
+func TestMerge_TypeNumericConflictResolved(t *testing.T) {
+	merged, err := Merge(Schema{
+		AllOf: SchemaRefs{
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "integer",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "number",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, "integer", merged.Properties["prop1"].Value.Type)
+}
+
+// Conflicting types cannot be resolved
+func TestMerge_TypeFailure(t *testing.T) {
+	_, err := Merge(Schema{
+		AllOf: SchemaRefs{
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "integer",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Properties: Schemas{
+						"prop1": &SchemaRef{
+							Value: &Schema{
+								Type: "string",
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+		},
+	})
+
+	require.EqualError(t, err, TypeErrorMessage)
```

Comment:
This error is used for two different scenarios.
I recommend to use a dedicated error here.